src/stableswap/math.rs

```diff
@@ -0,0 +1,409 @@
+        );
+    }
+
+    #[test]
+    fn test_y_given_in() {
+        let precision = 1_u128;
+        let reserves = [1000u128, 2000u128];
+        let ann = 4u128;
+
+        let amount_in = 100u128;
+        assert_eq!(calculate_d(&reserves, ann, precision), Some(2942u128));
+        assert_eq!(
+            calculate_y_given_in(amount_in, reserves[0], reserves[1], ann, precision),
+            Some(2000u128 - 121u128)
+        );
+        assert_eq!(
+            calculate_d(&[1100u128, 2000u128 - 125u128], ann, precision),
+            Some(2942u128)
+        );
+    }
+
+    #[test]
+    fn test_y_given_out() {
+        let precision = 1_u128;
+        let reserves = [1000u128, 2000u128];
+        let ann = 4u128;
+
+        let amount_out = 100u128;
+
+        let expected_in = 83u128;
+
+        assert_eq!(calculate_d(&reserves, ann, precision), Some(2942u128));
+
+        assert_eq!(
+            calculate_y_given_out(amount_out, reserves[0], reserves[1], ann, precision),
+            Some(1000u128 + expected_in)
+        );
+        assert_eq!(
+            calculate_d(&[1000u128 + expected_in, 2000u128 - amount_out], ann, precision),
+            Some(2946u128)
+        );
+    }
+
+    #[test]
+    fn test_d_case() {
+        let amp = 400u128;
+        let ann = amp * 4u128;
+
+        let precision = 1u128;
+
+        let result = calculate_d(&[500000000000008580273458u128, 10u128], ann, precision);
+
+        assert!(result.is_some());
+    }
+
+    #[test]
+    fn test_d_case2() {
+        let amp = 168u128;
+        let ann = amp * 4u128;
+
+        let precision = 1u128;
+
+        let result = calculate_d(&[500000000000000000000010u128, 11u128], ann, precision);
+
+        assert!(result.is_some());
+    }
+}
+
+#[cfg(test)]
+mod invariants {
+    use super::two_asset_pool_math::*;
+    use super::{calculate_in_given_out, calculate_out_given_in};
+    use super::Balance;
+    use proptest::prelude::*;
+    use proptest::proptest;
+
+    pub const ONE: Balance = 1_000_000_000_000;
+
+    const RESERVE_RANGE: (Balance, Balance) = (100_000 * ONE, 100_000_000 * ONE);
+    const LOW_RESERVE_RANGE: (Balance, Balance) = (10_u128, 11_u128);
+    const HIGH_RESERVE_RANGE: (Balance, Balance) = (500_000_000_000 * ONE, 500_000_000_001 * ONE);
+
+    fn trade_amount() -> impl Strategy<Value = Balance> {
+        //Just(1000 * ONE)
+        1000..10000 * ONE
+    }
+
+    fn high_trade_amount() -> impl Strategy<Value = Balance> {
+        500_000_000_000 * ONE..500_000_000_001 * ONE
+    }
+
+    fn asset_reserve() -> impl Strategy<Value = Balance> {
+        RESERVE_RANGE.0..RESERVE_RANGE.1
+    }
+    fn low_asset_reserve() -> impl Strategy<Value = Balance> {
+        LOW_RESERVE_RANGE.0..LOW_RESERVE_RANGE.1
+    }
+    fn high_asset_reserve() -> impl Strategy<Value = Balance> {
+        HIGH_RESERVE_RANGE.0..HIGH_RESERVE_RANGE.1
+    }
+
+    fn amplification() -> impl Strategy<Value = Balance> {
+        5..10000u128
```

Comment:
Is this correct? Should we get higher number for stableswap?

Reply:
sure. we can set it to any range.  

any suggestions on what max value can be ?

Reply:
cc @lumir-mrkva @poliwop

Reply:
stableswap pallet will have min/max amplification range. 

which was decided to be currently 2..10000.  It will be used here  too.